engine/collection/epochmgr/engine.go

```diff
@@ -537,11 +558,36 @@ func (e *Engine) removeEpoch(counter uint64) {
 // No errors are expected during normal operation.
 func (e *Engine) activeClusterIDs() (flow.ChainIDList, error) {
 	e.mu.RLock()
-	defer e.mu.RUnlock()
-	clusterIDs := make(flow.ChainIDList, 0)
-	for _, epoch := range e.epochs {
+	clusterIDs := make(flow.ChainIDList, len(e.epochs))
+	for i, epoch := range e.epochs {
 		chainID := epoch.state.Params().ChainID() // cached, does not hit database
-		clusterIDs = append(clusterIDs, chainID)
+		clusterIDs[i] = chainID
 	}
+	e.mu.RUnlock()
 	return clusterIDs, nil
 }
+
+// trackInProgressVote stores the context for an in progress vote process in the in progress vote map.
+func (e *Engine) trackInProgressVote(id string, cancel context.CancelFunc) {
+	e.qcVoteMu.Lock()
+	e.inProgressQCVotes[id] = cancel
+	e.qcVoteMu.Unlock()
+}
+
+// removeInProgressVote removes the context for an in progress vote process from the in progress vote map.
+func (e *Engine) removeInProgressVote(id string) {
+	e.qcVoteMu.Lock()
+	delete(e.inProgressQCVotes, id)
+	e.qcVoteMu.Unlock()
+}
+
+// stopAllInProgressQcVotes cancels the context for all in progress root qc voting.
+func (e *Engine) stopAllInProgressQcVotes() {
+	e.qcVoteMu.Lock()
+	for processId, cancel := range e.inProgressQCVotes {
+		cancel()
+		delete(e.inProgressQCVotes, processId)
+		e.log.Warn().Msgf("root qc vote process canceled: %s", processId)
+	}
+	e.qcVoteMu.Unlock()
+}
```

Comment:
with my suggestion to use atomics, this function could look like: 
```suggestion
// stopAllInProgressQcVotes cancels the context for the root QC voting (if there is any currently ongoing).
func (e *Engine) stopAllInProgressQcVotes() {
	for {
		c := e.inProgressQCVote.Load()
		if c != nil {
			e.log.Warn().Msgf("voting for cluster root block cancelled")
			(*c)() // cancel
		}
		if e.inProgressQCVote.CompareAndSwap(c, nil) {
			break
		}
	}
}
```